src/unix/process.c

```diff
@@ -343,6 +346,423 @@ static void uv__process_child_init(const uv_process_options_t* options,
+    dlsym(RTLD_DEFAULT, "posix_spawnattr_set_uid_np");
+  posix_spawn_fncs.spawnattr.set_gid_np = 
+    dlsym(RTLD_DEFAULT, "posix_spawnattr_set_gid_np");
+  posix_spawn_fncs.spawnattr.set_groups_np = 
+    dlsym(RTLD_DEFAULT, "posix_spawnattr_set_groups_np");
+  posix_spawn_fncs.file_actions.addchdir_np = 
+    dlsym(RTLD_DEFAULT, "posix_spawn_file_actions_addchdir_np");
+}
+
+
+int uv__spawn_set_posix_spawn_attrs(posix_spawnattr_t* attrs,
+                                    const uv__posix_spawn_fncs_t* posix_spawn_fncs,
+                                    const uv_process_options_t* options) {
+  int err;
+  unsigned int flags;
+  sigset_t signal_set;
+
+  err = posix_spawnattr_init(attrs);
+  if (err != 0) {
+    /* If initialization fails, no need to de-init, just return */
+    return err;
+  }
+
+  if (options->flags & UV_PROCESS_SETUID) {
+    if (posix_spawn_fncs->spawnattr.set_uid_np == NULL) {
+      err = ENOSYS;
+      goto error;
+    }
+
+    err = posix_spawn_fncs->spawnattr.set_uid_np(attrs, options->uid);
+    if (err != 0)
+      goto error;
+  }
+
+  if (options->flags & UV_PROCESS_SETGID) {
+    if (posix_spawn_fncs->spawnattr.set_gid_np == NULL) {
+      err = ENOSYS;
+      goto error;
+    }
+
+    err = posix_spawn_fncs->spawnattr.set_gid_np(attrs, options->gid);
+    if (err != 0) 
+      goto error;
+  }
+
+  if (options->flags & (UV_PROCESS_SETUID | UV_PROCESS_SETGID)) {
+    /* Using ngroups = 0 implied the group_array is empty, and so 
+     * its contents are never traversed. Still the 
+     * posix_spawn_set_groups_np function seems to require that the 
+     * group_array pointer be non-null */
+    const int ngroups = 0;
+    gid_t group_array = KAUTH_GID_NONE;
+
+    if (posix_spawn_fncs->spawnattr.set_groups_np == NULL) {
+      err = ENOSYS;
+      goto error;
+    }
+    
+    /* See the comment on the call to setgroups in uv__process_child_init above
+     * for why this is not a fatal error */
+    SAVE_ERRNO(posix_spawn_fncs->spawnattr.set_groups_np(
+      attrs, 
+      ngroups, 
+      &group_array, 
+      KAUTH_UID_NONE));
+  }
+
+  /* Set flags for spawn behavior 
+    * 1) POSIX_SPAWN_CLOEXEC_DEFAULT: (Apple Extension) All descriptors in
+    * the parent will be treated as if they had been created with O_CLOEXEC.
+    * The only fds that will be passed on to the child are those manipulated
+    * by the file actions
+    * 2) POSIX_SPAWN_SETSIGDEF: Signals mentioned in spawn-sigdefault in
+    * the spawn attributes will be reset to behave as their default
+    * 3) POSIX_SPAWN_SETSIGMASK: Signal mask will be set to the value of
+    * spawn-sigmask in attributes
+    * 4) POSIX_SPAWN_SETSID: Make the process a new session leader if a
+    * detached session was requested. */
+  flags = POSIX_SPAWN_CLOEXEC_DEFAULT |
+          POSIX_SPAWN_SETSIGDEF |
+          POSIX_SPAWN_SETSIGMASK;
+  if (options->flags & UV_PROCESS_DETACHED) 
+    flags |= POSIX_SPAWN_SETSID;
```

Comment:
This appears to have been unavailable until 10.15, and so will need to be defined manually and checked for at runtime

Reply:
Could you advise as what you expect in runtime? Conditionally executing with `__builtin_available` is not allowed, as far as I understand.  At _compile_ time, I could wrap this logic in `#ifndef POSIX_SPAWN_SETSID` and have the else side return `ENOSYS` to fall back to the fork flow, resolving builds done on 10.14. But I'm not sure how to runtime-check this one.

Reply:
This one is problematic. You can use the constant literal value (i.e. `1024 /* POSIX_SPAWN_SETSID */`) but the xnu kernel doesn't report errors for flags it doesn't recognize - posix_spawn silently spawns the process without calling setsid.

There's a mach system call you can use to find out the kernel version. It's the same output as `uname -v`.
```c
#include <mach/mach_host.h>
#include <stdio.h>
int main(void) {
  kernel_version_t version = {0};
  host_kernel_version(mach_host_self(), version);
  puts(version);
}
```
On my system prints:
```
Darwin Kernel Version 18.7.0: Thu Jun 20 18:42:21 PDT 2019; root:xnu-4903.270.47~4/RELEASE_X86_64
```

Reply:
You can check that with `sysctlbyname("kern.osrelease", str, &size, NULL, 0);` >= `19.0.0`

https://stackoverflow.com/questions/11072804/how-do-i-determine-the-os-version-at-runtime-in-os-x-or-ios-without-using-gesta

Reply:
I have implemented the `sysctlbyname` thing, and it seems to work just fine. We are running some tests on 10.14 after we though we had ran into some runtime issues, but I think I'll push the change very soon.

Reply:
Ok, pushed.